### PR TITLE
Landscaper control plane component - add import configuration and validation

### PR DESCRIPTION
**Pull request by danielfoehrKn**
**How to categorize this PR?**

/area control-plane
/kind enhancement

**What this PR does / why we need it**:

Adds the
- import configuration (with each control plane component having its own configuration in a dedicated file & commit)
   - the Gardener API server
   - the GCM
   - the Gardener Admission Controller
   - the Gardener Scheduler
- conversion coding
- validation coding
- README explaining the configuration values (will be expanded in the future)

of the Landscaper control-plane component.
**The goal is to create the component incrementally to expedite and ease the review process.** 

Essentially the component will apply the Gardener control plane chart (charts/gardener/controlplane).
The config file of the Landscaper control-plane component contains (with minor differences) the same configuration.
Adds type-safety and validation.
 
 In a subsequent PR,  I will add the actual application code.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```other operator
Add API types for the landscaper control plane component.
```

### CODE REVIEW

File: landscaper/pkg/controlplane/apis/imports/types_apiserver.go [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+
+// APIServerDeploymentConfiguration contains certain configurations for the deployment
+// of the Gardener Extension API server
+type APIServerDeploymentConfiguration struct {
+	// CommonDeploymentConfiguration contains common deployment configurations
+	// Defaults:
+	//  Resources: Requests (CPU: 100m, memory 100Mi), Limits (CPU: 300m, memory: 256Mi)
+	CommonDeploymentConfiguration
+	// LivenessProbe allows to overwrite the default liveness probe
+	// Defaults:
```
Comment (schrodit) [inline landscaper/pkg/controlplane/apis/imports/types_apiserver.go:45]: I do not see any defaulting here. Is the defaulting added in the external type or do you plan to add it in the helm chart?

Comment (danielfoehrKn) [reply landscaper/pkg/controlplane/apis/imports/types_apiserver.go:45]: I will add the defaulting in the external type. 
However, I am planning not to duplicate defaulting, which is already done in the helm chart.
At least that is currently the plan, but not 100% sure if this is how it should be done.

WDYT also @rfranzke

**Comment by danielfoehrKn** [comment]
@rfranzke @schrodit or anyone else wants to be the main contact person for the control-plane component story? 
Will be a sequence of PRs where it would help to have an involved reviewer.

**Comment by rfranzke** [comment]
I will definitely look into it, though, this week is KubeCon and a few other large PRs are ongoing in parallel.

**Comment by schrodit** [comment]
> @rfranzke @schrodit or anyone else wants to be the main contact person for the control-plane component story?
> Will be a sequence of PRs where it would help to have an involved reviewer.

I will also have a look. Specifically on the Landscaper part.

File: landscaper/pkg/controlplane/apis/imports/types_apiserver.go (hunk 1 above)
Comment (schrodit) [reply landscaper/pkg/controlplane/apis/imports/types_apiserver.go:45]: > However, I am planning not to duplicate defaulting

Yes that was the background of my question. I think defaulting in different places will result in confusion and could be hard to debug.

File: charts/gardener/controlplane/local-values.yaml [hunk 2]
```diff
@@ -3,6 +3,7 @@
 global:
   # Gardener API server configuration values
   apiserver:
+    clusterIdentity: garden-cluster-identity
```
Comment (rfranzke) [inline charts/gardener/controlplane/local-values.yaml]: See below - for the `local-values.yaml` it probably doesn't matter too much. Actually, what is the `local-values.yaml` used for? Can it be deleted?

File: charts/gardener/controlplane/values.yaml [hunk 3]
```diff
@@ -1,7 +1,10 @@
+# NOTE: When making changes to default values & validation, please adjust the API comments and validation
+# in package landscaper/pkg/controlplane/apis/imports
 global:
   # Gardener API server configuration values
   apiserver:
     enabled: true
+    clusterIdentity: garden-cluster-identity
```
Comment (rfranzke) [inline charts/gardener/controlplane/values.yaml]: This value was consciously not added to the `values.yaml` to force operators to provide a dedicated name. Otherwise, it can easily lead to conflicts and strange behaviour. Please remove it again.

File: charts/gardener/controlplane/values.yaml [hunk 4]
```diff
@@ -405,10 +409,6 @@ global:
           bindAddress: 0.0.0.0
           port: 2719
           tls:
-            caBundle: |
-              -----BEGIN CERTIFICATE-----
-              ...
-              -----END CERTIFICATE-----
```
Comment (rfranzke) [inline charts/gardener/controlplane/values.yaml:415]: Why was this removed?

File: landscaper/pkg/controlplane/apis/imports/types_imports.go [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+// Prerequisites:
+//  - existing etcd cluster that is accessible from the runtime cluster
+//  - there must be a Kubernetes API server that is properly set up for API aggregation
+//  - The kubeconfig for this API server has to be configured either in the field RuntimeCluster or
+//    in the field VirtualGarden in the import configuration
+// This component extends the given Kubernetes API server with the Gardener API server that serves
+// all of the Gardener resource groups (e.g for Shoots)
+// To setup API server aggregation, the Kubernetes API server has to be properly configured with configuration flags.
+// For more information how to configure the aggregation layer see:
+// https://kubernetes.io/docs/tasks/extend-kubernetes/configure-aggregation-layer
```
Comment (rfranzke) [inline landscaper/pkg/controlplane/apis/imports/types_imports.go]: Hm, why is this here? To me it sounds to better fit into a document and not into a comment inside the code?

File: landscaper/pkg/controlplane/apis/imports/types_admission_controller.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+	// can be configured here
+	ComponentConfiguration runtime.Object
+}
+
+// SeedRestriction configures the SeedRestriction admission plugin
+type SeedRestriction struct {
+	// Enabled configures whether the SeedRestriction admission plugin should be enabled.
+	// Sets up the ValidatingWebhookConfiguration pointing to the webhook server in the Gardener Admission Controller
+	// serving the SeedRestriction webhook
+	// Default: true
```
Comment (rfranzke) [inline landscaper/pkg/controlplane/apis/imports/types_admission_controller.go]: Shouldn't it be a `*bool` then? Otherwise, how do you know when to apply the default value?

File: landscaper/pkg/controlplane/apis/imports/types_apiserver.go (hunk 1 above)
Comment (rfranzke) [reply landscaper/pkg/controlplane/apis/imports/types_apiserver.go:45]: Good point, how was it done for the gardenlet landscaper component?

**Review by rfranzke — COMMENTED** [review]
What about unit tests for the validation logic?

File: charts/gardener/controlplane/values.yaml (hunk 4 above)
Comment (danielfoehrKn) [reply charts/gardener/controlplane/values.yaml:415]: I checked and did not find a usage for the GCM for the CA bundle. 
Am I missing something?

File: charts/gardener/controlplane/values.yaml (hunk 3 above)
Comment (danielfoehrKn) [reply charts/gardener/controlplane/values.yaml]: okay I see, will do.

File: charts/gardener/controlplane/local-values.yaml (hunk 2 above)
Comment (danielfoehrKn) [reply charts/gardener/controlplane/local-values.yaml]: I am not sure. I think the idea was to provide a minimal viable values.yaml with some standard configuration for local testing. 
But I never used it or introduced it.
I would say it can be deleted.

File: landscaper/pkg/controlplane/apis/imports/types_imports.go (hunk 5 above)
Comment (danielfoehrKn) [reply landscaper/pkg/controlplane/apis/imports/types_imports.go]: I will also add a more detailed version to the documentation for the component itself. 
I think that is helpful information to understand the configuration options. I would like to keep it.

**Comment by danielfoehrKn** [comment]
> What about unit tests for the validation logic?

I first wanted to verify the API itself to avoid double maintenance --> having to adjust both the validation and its unit tests. 
I will add it in the next PR together with the external API version & defaulting to complete the `api` package. 
Does that sound ok?

File: landscaper/pkg/controlplane/apis/imports/types_apiserver.go (hunk 1 above)
Comment (danielfoehrKn) [reply landscaper/pkg/controlplane/apis/imports/types_apiserver.go:45]: defaulting was not duplicated.

In general, I wanted to make the deployment as easy as possible.
If there is something reasonable to be defaulted / generated which is currently not (e.g the TLS certificates and CA Bundles) then I'll add that. 
- given that there is no existing CA bundle deployed yet

Also for hvpa, if the user does not explicitly configure it,  I want to only enable it if the hvpa CRD exists. Otherwise hvpa will not be enabled per default.

**Comment by danielfoehrKn** [comment]
Picking up this PR again.

File: charts/gardener/controlplane/values.yaml (hunk 3 above)
Comment (danielfoehrKn) [reply charts/gardener/controlplane/values.yaml]: I would propose to generate a random gardener identity in the controlplane component if not explicitly provided (making deployment easy).
I will remove the value from the helm chart though as suggested.

File: charts/gardener/controlplane/local-values.yaml (hunk 2 above)
Comment (danielfoehrKn) [reply charts/gardener/controlplane/local-values.yaml]: I would delete it in a separate PR. This PR does not touch the existing loca-values.yaml

File: landscaper/pkg/controlplane/apis/imports/types_imports.go (hunk 5 above)
Comment (danielfoehrKn) [reply landscaper/pkg/controlplane/apis/imports/types_imports.go]: I added a dedicated README file with more detailed information and removed this comment.

File: charts/gardener/controlplane/values.yaml [hunk 7]
```diff
@@ -253,8 +255,6 @@ global:
  #      truncateMaxBatchSize: 10485760
  #      truncateMaxEventSize: 102400
  #      version: audit.k8s.io/v1
- #  clusterIdentity: garden-cluster-identity
-
```
Comment (rfranzke) [inline charts/gardener/controlplane/values.yaml:257]: Why was this removed?

File: pkg/scheduler/apis/config/encoding/encoding.go [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+		return nil, err
+	}
+	return &runtime.RawExtension{
+		Raw:    raw,
+		Object: cfg,
+	}, nil
+}
+
+// EncodeschedulerConfigurationToBytes encodes the given external schedulerConfiguration version into a byte slice.
+func EncodeschedulerConfigurationToBytes(cfg *schedulerconfigv1alpha1.SchedulerConfiguration) ([]byte, error) {
```
Comment (rfranzke) [inline pkg/scheduler/apis/config/encoding/encoding.go]: use camel case

File: pkg/scheduler/apis/config/encoding/encoding.go [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+		},
+	}
+	if _, _, err := getDecoder(withDefaults).Decode(bytes, nil, cfg); err != nil {
+		return nil, err
+	}
+	return cfg, nil
+}
+
+// EncodeschedulerConfiguration encodes the given external GardenletConfiguration version into a raw extension.
+func EncodeschedulerConfiguration(cfg *schedulerconfigv1alpha1.SchedulerConfiguration) (*runtime.RawExtension, error) {
```
Comment (rfranzke) [inline pkg/scheduler/apis/config/encoding/encoding.go]: use camel case

File: pkg/scheduler/apis/config/helper/helper_suite_test.go [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+import (
+	"testing"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+func TestHelper(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "Gardener Scheduler API Helper Suite")
```
Comment (rfranzke) [inline pkg/scheduler/apis/config/helper/helper_suite_test.go]:
```suggestion
	RunSpecs(t, "Gardener Scheduler APIs Config Helper Suite")
```

File: pkg/admissioncontroller/apis/config/encoding/encoding_suite_test.go [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+import (
+	"testing"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+func TestEncoding(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "ControllerManager API Encoding Suite")
```
Comment (rfranzke) [inline pkg/admissioncontroller/apis/config/encoding/encoding_suite_test.go]:
```suggestion
	RunSpecs(t, "AdmissonController API Encoding Suite")
```

File: charts/gardener/controlplane/values.yaml [hunk 12]
```diff
@@ -5,7 +5,7 @@ global:
   apiserver:
     enabled: true
     serviceEnabled: true
-    replicaCount: 1
+    replicaCount: 3
```
Comment (rfranzke) [inline charts/gardener/controlplane/values.yaml]: Why is this changed?

File: landscaper/pkg/gardenlet/controller/gardenlet_delete.go [hunk 13]
```diff
@@ -49,11 +49,11 @@ func (g Landscaper) Delete(ctx context.Context) error {
...
 			return fmt.Errorf("cannot delete seed '%s' which is still used by at least one shoot", g.gardenletConfiguration.SeedConfig.Name)
 		}
+	}
 
-		err := g.deleteSeedAndWait(ctx, seed)
-		if err != nil {
-			return err
-		}
+	err = g.deleteSeedAndWait(ctx, seed)
+	if err != nil {
```
Comment (rfranzke) [inline landscaper/pkg/gardenlet/controller/gardenlet_delete.go]: Why this change?/why not this:

```suggestion
	if err := g.deleteSeedAndWait(ctx, seed); err != nil {
```

File: landscaper/pkg/controlplane/apis/imports/types_imports.go [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+}
+
+// VirtualGarden contains configuration for the "Virtual Garden" setup option of Gardener
+type VirtualGarden struct {
+	// Enabled configures whether to setup Gardener with the "Virtual Garden" setup option of Gardener
+	// Please note that as a prerequisite, the API server pods of the "Virtual Garden" already need to be deployed to
+	// the runtime cluster (this should be done automatically by a preceding component when using the standard installation via the landscaper)
+	// and must be able to communicate with the Gardener Extension API server pod that will
+	// be deployed to the Garden namespace
+	Enabled *bool
```
Comment (rfranzke) [inline landscaper/pkg/controlplane/apis/imports/types_imports.go]: Why this is a pointer?

File: landscaper/pkg/controlplane/apis/imports/v1alpha1/types_admission_controller.go [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+	// +optional
+	*Configuration `json:"inline,omitempty"`
+}
+
+// SeedRestriction configures the SeedRestriction admission plugin
+type SeedRestriction struct {
+	// Enabled configures whether the SeedRestriction admission plugin should be enabled.
+	// Sets up the ValidatingWebhookConfiguration pointing to the webhook server in the Gardener Admission Controller
+	// serving the SeedRestriction webhook
+	// Default: true
```
Comment (rfranzke) [inline landscaper/pkg/controlplane/apis/imports/v1alpha1/types_admission_controller.go]: The default should be false, I guess?

File: landscaper/pkg/controlplane/README.md [hunk 16]
```diff
@@ -0,0 +1,441 @@
...
+- you can directly **configure and upgrade the API server** itself (e.g. to set up API aggregation) - **this is typically not the case on managed Kubernetes offerings**.
+- there are no performance concerns considering the `runtime` cluster's API server will serve both the Gardener API plus hosting other workload. 
+
+### 2) Virtual Garden
+
+**Prerequisites**
+- An Kubernetes API server deployment configured for API aggregation in the `runtime` cluster.
+  This component does not deploy the such API server. It must already exist in the runtime cluster
+  and setup for [API aggregation](https://kubernetes.io/docs/tasks/extend-kubernetes/configure-aggregation-layer).
+  This is the case when the [virtual garden component](https://github.com/gardener/virtual-garden) of the lanscaper has already been deployed to the `runtime` cluster.
```
Comment (rfranzke) [inline landscaper/pkg/controlplane/README.md]:
```suggestion
  This is the case when the [virtual garden component](https://github.com/gardener/virtual-garden) of the landscaper has already been deployed to the `runtime` cluster.
```

File: landscaper/pkg/controlplane/README.md [hunk 17]
```diff
@@ -0,0 +1,441 @@
...
+
+In this setup,
+- the `runtime` cluster hosts the Gardener control Plane pods
```
Comment (rfranzke) [inline landscaper/pkg/controlplane/README.md]: s/Plane/plane

File: landscaper/pkg/controlplane/README.md [hunk 18]
```diff
@@ -0,0 +1,441 @@
...
+- a dedicated Kubernetes API server deployed in the `runtime` cluster is extendet by the Gardener Extension API server (`virtual-garden` API server).
```
Comment (rfranzke) [inline landscaper/pkg/controlplane/README.md]: s/extendet/extended

**Review by rfranzke — COMMENTED** [review]
Generally it looks OK, I only have minor comments. What about unit tests for the validation functions?

File: landscaper/pkg/controlplane/apis/imports/v1alpha1/types_admission_controller.go (hunk 15 above)
Comment (danielfoehrKn) [reply landscaper/pkg/controlplane/apis/imports/v1alpha1/types_admission_controller.go]: Secure by default? :)

Comment (rfranzke) [reply landscaper/pkg/controlplane/apis/imports/v1alpha1/types_admission_controller.go]: While the intention is great, there are still some manual steps that need to be performed before this can be set to `true`, see https://github.com/gardener/gardener/blob/master/docs/deployment/gardenlet_api_access.md#seedrestriction-admission-webhook-enablement.

File: charts/gardener/controlplane/values.yaml (hunk 12 above)
Comment (danielfoehrKn) [reply charts/gardener/controlplane/values.yaml]: The Admission Controller gets [3 replicas per default](https://github.com/gardener/gardener/blob/ed03f939949dfd80ec7317ab72f361dfff05dfc0/charts/gardener/controlplane/values.yaml#L261), this does not match the one replica for the API server getting most of the load. I thought it is a sensible default. But I can surely remove it again.

File: charts/gardener/controlplane/values.yaml (hunk 7 above)
Comment (danielfoehrKn) [reply charts/gardener/controlplane/values.yaml:257]: It seems like I added that in an earlier commit in the series and then removed it again based on your previous suggestion.

File: landscaper/pkg/controlplane/apis/imports/v1alpha1/types_admission_controller.go (hunk 15 above)
Comment (danielfoehrKn) [reply landscaper/pkg/controlplane/apis/imports/v1alpha1/types_admission_controller.go]: okay, so only if the Seed Authorizer is enabled, this can be true. Shall it be defaulted to true in the case the Seed Authorizer is enabled?

File: charts/gardener/controlplane/values.yaml (hunk 12 above)
Comment (rfranzke) [reply charts/gardener/controlplane/values.yaml]: Yeah, let's keep it `1` for now. Probably we could make the default for GAC `1` as well to not waste too many resources in small dev/experimental landscapes.

File: landscaper/pkg/controlplane/apis/imports/v1alpha1/types_admission_controller.go (hunk 15 above)
Comment (rfranzke) [reply landscaper/pkg/controlplane/apis/imports/v1alpha1/types_admission_controller.go]: Yes, good idea, I think this makes sense!

File: landscaper/pkg/gardenlet/controller/gardenlet_delete.go (hunk 13 above)
Comment (danielfoehrKn) [reply landscaper/pkg/gardenlet/controller/gardenlet_delete.go]: Reverted the change in error. thanks for spotting.

File: landscaper/pkg/controlplane/apis/imports/v1alpha1/types_admission_controller.go (hunk 15 above)
Comment (danielfoehrKn) [reply landscaper/pkg/controlplane/apis/imports/v1alpha1/types_admission_controller.go]: Added the defaulting.

**Review by rfranzke — APPROVED** [review]
/lgtm
